File: LeetCode/LeetCode 1262. 可被三整除的最大和 (动态规划).py
class Solution:
    def maxSumDivThree(self, nums: List[int]) -> int:
        dp = [0, -1e9, -1e9]
        
        for x in nums:
            mod = x % 3
            now = [0, 0, 0]
            for i in range(3):
                now[i] = max(dp[i], dp[(3 + i - mod) % 3] + x)
            dp = now
        
        return dp[0]

File: LeetCode/LeetCode 1262. 可被三整除的最大和 (动态规划).cpp
//dp[i]:表示当前选择的数字累加模3后的最大数字和
class Solution {
public:
    int maxSumDivThree(vector<int>& nums) {
        vector dp = {0, INT_MIN, INT_MIN};
        for(auto &x : nums) {
            int mod = x % 3;
            vector <int> now(3, 0);
            for(int i = 0; i < 3; i++) {
                now[i] = max(dp[i], dp[(3 + i - mod) % 3] + x);
            }
            dp = now;
        }
        return dp[0];
    }
};